app/src/main/java/org/oppia/android/app/home/promotedlist/PromotedStoryListView.kt

```diff
@@ -47,12 +48,20 @@ class PromotedStoryListView @JvmOverloads constructor(
    *
    * @param newDataList the new list of stories to present
    */
-  fun setDataList(newDataList: List<PromotedStoryViewModel>) {
-    // Update the adapter and the story list only if the list is new. The parent presenter should
-    // not render promoted stories if the list is empty, but default to showing the last list.
-    if (newDataList != null && !newDataList.isEmpty() && newDataList != dataList) {
-      dataList = newDataList
+  fun setPromotedStoryList(@Nullable newDataList: List<PromotedStoryViewModel>) {
+    // To reliably bind data only after the adapter is created, we manually set the data so we can first
+    // check for the adapter; when using an existing [RecyclerViewBindingAdapter] there is no reliable
+    // way to check that the adapter is created.
+    // This ensures that the adapter will only be created once and correctly rebinds the data.
+    // For more context: https://github.com/oppia/oppia-android/pull/2246#pullrequestreview-565964462
+    if (adapter == null) {
       adapter = createAdapter()
+    }
+
+    if (newDataList == null || newDataList.isEmpty()) {
```

Comment:
Still confused about this--why is the empty check needed? That seems like a potential legitimate state for the list to be in.

(Also for future reference, you can use ``newDataList.isNullOrEmpty()`` which is a convenient Kotlin function for cases like this).

Reply:
If the list is empty then we wouldn't want to bind it, I believe we wouldn't display any promoted stories but then the header is still visible.

Reply:
That's true (that the header would still be displayed), but an empty list doesn't necessarily mean error. If we don't want to show the header on an empty list, we need to handle that separately (and at the same time shouldn't show the recycler view, either). The problem I potentially see here is when we go from some data to no data in a legitimate case: we would still show cards even though we shouldn't anymore since the data list is now empty. Given that the data list is immutable that's technically not possible, but as the code changes in the future it may be made to be dynamic.

Reply:
Got it -- I'll remove the empty case then. As an FYI I think we do handle that case if the list is empty since in `HomeViewModel` we don't add a `PromotedStoryListViewModel` if the list is empty.

Reply:
Done

Reply:
Sounds good, though it seems you also removed the null case but I started another thread for that. Closing this one.